### PR TITLE
Introduce option to enable trim analysis warnings

### PR DESCRIPTION
**Pull request by sbomer**
New versions of the linker will emit warnings for code that can be broken by trimming. https://github.com/mono/linker/pull/1324 ~~will add~~ added support for `--nowarn`, which we will use to disable them (the "opinionated default") in .NET5 to prevent noise. This change turns them off and adds a new property to let developers opt into the warnings.

The new property is called `SuppressTrimAnalysisWarnings`.

Old notes below:
<hr />

I have called the new option `TrimAnalysis`. @samsp-msft, we would love your feedback on the option name.
 Other options considered:
- `TrimCorrectnessAnalysis`
- `TrimmerCorrectnessAnalysisWarnings`
- `ExperimentalTrimAnalysis`
- `EnableTrimAnalysis`

(let me know if I forgot any @vitek-karas @mateoatr)

I find "Correctness" a bit redundant since the warnings are all about different aspects of correctness. I left out the "Warnings" suffix because I think of this like enabling an analyzer (even though at the implementation level, in this particular case the analysis code always runs). I left out the "Enable" prefix because I think it makes sense for it to be a noun like the feature switches (e.g. `InvariantGlobalization`).

See https://github.com/mono/linker/pull/1303#issuecomment-653433498 for some more discussion of the approach.

Depends on https://github.com/mono/linker/pull/1324 which is a WIP, but we wanted to reach agreement on the naming in parallel.

### CODE REVIEW

**Comment by samsp-msft** [comment]
Is TrimAnalysis the name of the category, or the name of an option? What is an example of the command line where/how --nowarn and/or it are used. How does the user select different levels of trimming (Assembly, Type, Members)? As the naming should all go together.

**Comment by sbomer** [comment]
`TrimAnalysis` is the name of an option. We might also make it the name of a category, but that isn't exposed to the user yet (please correct me if your understanding is different @mateoatr @vitek-karas). It can be set in MSBuild as:
```xml
<PropertyGroup>
  <TrimAnalysis>true</TrimAnalysis>
</PropertyGroup>
```

When `TrimAnalysis` is *not* set, the linker is called with `dotnet illink.dll --nowarn IL2006;IL2026;IL2043`. (Technically the linker will get all `NoWarn`s passed to it, but only those with an `IL` prefix have any effect.)

When `TrimAnalysis` *is* set, the user can still manually set `<NoWarn>$(NoWarn);IL2006</NoWarn>` for example.

The existing option to set the "level" of trimming (we don't have one for type-level, and the assembly-level one might not do exactly what you expect) can be set like this:
```xml
<PropertyGroup>
  <TrimMode>link</TrimMode>
</PropertyGroup>
```

**Comment by mateoatr** [comment]
> `TrimAnalysis` is the name of an option. We might also make it the name of a category, but that isn't exposed to the user yet 

After https://github.com/mono/linker/pull/1324 is merged, the subcategory will be exposed to the users in the warning messages, e.g.: `ILLink: Trim analysis warning IL2006: <some warning>`.

**Comment by samsp-msft** [comment]
If the option is just for the reporting of errors warnings, then I would say "ReportTrimAnalysis", to make it clear that its about the errors and not whether it actually trims.

**Comment by vitek-karas** [comment]
I think that Report and Analysis are in this case more or less the same thing. Just like Roslyn analyzers, the fact that I'm going to run the analysis automatically means the results are reported (otherwise there's no point in running it).

We could also model it directly by roslyn analyzers and call it `TrimAnalyzer` or `EnableTrimAnalyzer`. In the future where we do have both linker and roslyn analyzer for this, it could turn on both... (in fact whatever option we pick here should turn on both once we have them).

**Comment by sbomer** [comment]
Another thing to consider is that we may in the future add the ability to analyze for single-file compatibility, for example. In that case the option might actually influence the behavior (telling it to analyze everything, not just referenced code, for example). Maybe it's too early to design that though - and we can always add a new option that sets a default for `ReportTrimAnalysis`.

**Comment by sbomer** [comment]
> In the future where we do have both linker and roslyn analyzer for this, it could turn on both

As far as I know, the typical way to turn on a roslyn analyzer is by adding a packagereference. @agocke do you know of any precedent that would be relevant here (analyzers shipped with the SDK, maybe)?

**Comment by sbomer** [comment]
See https://github.com/mono/linker/issues/1367 for a discussion of the warnings we will control via this new option.

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs [hunk 1]
```diff
@@ -560,12 +664,13 @@ public void ILLink_displays_informational_warning(string targetFramework)
...
             var projectName = "HelloWorld";
             var referenceProjectName = "ClassLibForILLink";
+            var rid = EnvironmentInfo.GetCompatibleRid(targetFramework);
 
             var testProject = CreateTestProjectForILLinkTesting(targetFramework, projectName, referenceProjectName);
             var testAsset = _testAssetsManager.CreateTestProject(testProject);
 
             var publishCommand = new PublishCommand(Log, Path.Combine(testAsset.TestRoot, testProject.Name));
-            publishCommand.Execute("/p:PublishTrimmed=true", $"/p:SelfContained=true", "/p:PublishTrimmed=true")
+            publishCommand.Execute("/p:PublishTrimmed=true", $"/p:RuntimeIdentifier={rid}", "/p:SelfContained=true", "/p:PublishTrimmed=true")
```
Comment (eerhardt) [inline src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:673]: Why was this needed?

**Review by eerhardt — APPROVED** [review]
:shipit:

File: src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs (hunk 1 above)
Comment (sbomer) [reply src/Tests/Microsoft.NET.Publish.Tests/GivenThatWeWantToRunILLink.cs:673]: When I added `IsExe` to the test project (it should have been set to begin with), this errored because no RID was specified. It worked when it was a library project where SelfContained doesn't do anything (and the test doesn't actually check anything about the produced files - just looks for a warning message being printed). The other tests pass a RID like they should.

**Comment by sbomer** [comment]
Moved to https://github.com/dotnet/sdk/pull/12815 to do a fresh CI run. (I thought I would be able to re-open this one, but not so after a rebase).
